Automattic/wp-calypso
# Products: Remove some unnecessary uses of formatProduct

#### Changes proposed in this Pull Request

The `formatProduct` function in the `calypso-products` package is used superstitiously and hides the data types being passed to the functions that use it. Basically it takes an object and snake_cases several of its properties, but in so doing it becomes very difficult to know what the object is that's being manipulated and in many cases the formatting has no effect. This also makes it difficult to convert code to TypeScript.

As an example, there's a minor bug in the function `productMatches` which is uncovered and fixed by this PR.

This PR removes the use of `formatProduct` in the following functions, converting them all to TypeScript in the process.

- `getProductFromSlug` (already returns data in an explicit format)
- `getJetpackProductShortName` (all uses already have snake_case slug)
- `getJetpackProductDisplayName` (all uses already have snake_case slug)
- `getJetpackProductDescription` (all uses already have snake_case slug)
- `getJetpackProductCallToAction` (all uses already have snake_case slug)
- `getJetpackProductTagline` (normalized this function's arguments in the one location that needed it)
- `getIncludedDomainPurchaseAmount` (removed this function as it could be replaced inline at its one call site)
- `getDomainProductRanking` (removed this unused function)
- `includesProduct` (removed this unused function)

There are many more functions to alter but I felt that this was sufficient for one PR. https://github.com/Automattic/wp-calypso/pull/58086 removes some more.

Future work could involve changing these functions to just take a product slug instead (or whatever property they require) but as that would involve changing all call sites, it felt out-of-scope for this change.

#### Testing instructions

Verify that the call-sites of the altered functions all pass in their product argument with its relevant properties (mainly just `product_slug`) in snake_case format.
